Recover from crashing docker socket

achntrl opened the pull request:
### What does this PR do?

It restarts the docker tailers to reopen a connection to the docker socket

### Motivation

When the docker socket is killed, the log ingestion for container fails because the reader isn't properly closed

achntrl committed:
Restart a tailer on unexpected EOF on docker socket

achntrl committed:
Impletement a retry logic for docker tailer restart

achntrl committed:
Add release note

ajacquemot commented on pkg/logs/input/docker/tailer.go:
```diff
@@ -46,10 +46,11 @@ type Tailer struct {
 	shouldStop    bool
 	stop          chan struct{}
 	done          chan struct{}
+	unexpectedEOF chan string
```
I think this channel should be renamed into something more friendly for its consumer, it's misleading as we don't know what it contains (a `containerID`).

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -169,3 +175,45 @@ func (l *Launcher) stopTailer(containerID string) {
+			backoffDuration = backoffMax
+		}
+
+		time.Sleep(backoffDuration)
```
You sleep for all restarts here, you should not sleep during the first iteration.

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -30,6 +32,7 @@ type Launcher struct {
 	cli               *client.Client
 	registry          auditor.Registry
 	stop              chan struct{}
+	lostSocket        chan string
```
Same here, the . name of the channel is misleading and does not really reflect what it contains.

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -169,3 +175,45 @@ func (l *Launcher) stopTailer(containerID string) {
 }
+
+func (l *Launcher) restartTailer(containerID string) {
+	backoffDuration := 1 * time.Microsecond
```
I don't think that a `Microsecond` is a good reference here, I think we should use a `Second` instead, WDYT ?

achntrl replied:
Yes, I was actually using `Second` but I switch to `Microsecond` to manually test the retry logic (because the docker socket restarts really quickly)

ajacquemot reviewed (commented):
Thanks for your work, this is going to the right direction but I left few comments.
Also I don't think tailers need to use a channel to notify the launcher that they can't collect logs anymore, did you think about other solutions ?

achntrl committed:
Rename channels

achntrl committed:
Retry on all errors

achntrl commented:
@ajacquemot 
>Also I don't think tailers need to use a channel to notify the launcher that they can't collect logs anymore, did you think about other solutions ?

Why ? Isn't the channel the standard medium of communication between two goroutines ?

achntrl committed:
Fix race condition on agent close

mnshdw commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -21,27 +23,29 @@ import (
+		pendingContainers:  make(map[string]*Container),
+		registry:           registry,
+		stop:               make(chan struct{}),
+		erroredContainerID: make(chan string, 1),
```
Why limit to 1 ?

mnshdw commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -169,3 +175,45 @@ func (l *Launcher) stopTailer(containerID string) {
+
+		// compute the offset to prevent from missing or duplicating logs
+		since, err := Since(l.registry, tailer.Identifier(), service.Before)
+		if err != nil {
```
I understand the motivation behind fetching the offset before resuming but can this ever recover ?
If it fails once I think it is going to fail forever.

mnshdw replied:
We should probably use the same logic as other tailers, which I think is to log the error and tail from the end (which is likely what `Since` returns in case of error).

achntrl replied:
👍

mnshdw commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -169,3 +175,45 @@ func (l *Launcher) stopTailer(containerID string) {
+
+func (l *Launcher) restartTailer(containerID string) {
+	backoffDuration := 1 * time.Second
+	backoffMax := 60 * time.Second
```
Probably could be a const defined outside of this method.

mnshdw commented on pkg/logs/input/docker/tailer.go:
```diff
@@ -135,10 +145,16 @@ func (t *Tailer) readForever() {
 			n, err := t.reader.Read(inBuf)
 			if err != nil {
+				if isClosedConnError(err) {
+					log.Warn("Closed conn")
```
I'm not sure this log is really useful as-is.

mnshdw commented on pkg/logs/input/docker/tailer.go:
```diff
@@ -135,10 +145,16 @@ func (t *Tailer) readForever() {
 				// an error occurred, stop from reading new logs
 				if err != io.EOF {
 					t.source.Status.Error(err)
 					log.Error("Err: ", err)
```
Not your code, but I find this error message lacking context.

mnshdw commented on pkg/logs/input/docker/tailer.go:
```diff
@@ -135,10 +145,16 @@ func (t *Tailer) readForever() {
 				if err != io.EOF {
 					t.source.Status.Error(err)
 					log.Error("Err: ", err)
+					log.Warn("The tailer of container ", ShortContainerID(t.ContainerID), " will restart")
```
We could probably merge both messages to simplify a bit, eg. "Tailing container X stopped because of error Y"
Also logging the fact that we are (re)starting tailing should probably be done where we consume erroredContainerID.

mnshdw reviewed (commented):
Left a few minor comments.

achntrl committed:
Recover for failure on reading the registry

achntrl committed:
Change formatting of logs

achntrl committed:
Refactor readForever

achntrl committed:
Print short container name

achntrl committed:
Log container tailing restart

achntrl committed:
Fix a race condition

achntrl committed:
Update status when EOF is received after partition

When the network is partitioned, the tailing of a container is
restarting and the container is killed during the partition, the restart
function will tail the logs generated during the partition until it
reaches EOF (so we don't lose logs)
However, since the `tailer.Stop()` is not called, we wouldn't update the
status of the agent and the container would appear as if it was
still existing and tailed (even if `readForever()` exited and no
goroutine was working on this container). This commit update the status
message

achntrl committed:
Remove nil pointer error when stopping and restarting at the same time

achntrl committed:
Change backoff policy to expire after 60 sec

mnshdw commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -159,13 +173,68 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
+}
+
+func (l *Launcher) restartTailer(containerID string) {
+	log.Warnf("Tailing will restart for %v", ShortContainerID(containerID))
```
Not necessarily, depending on how it goes.
I think we should either include the notion that we are trying, or move/remove this log.

mnshdw commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -159,13 +173,68 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
+
+	for {
+		if backoffDuration > backoffMaxDuration {
+			log.Warnf("Could not restart container %v. Stop tailing", ShortContainerID(containerID))
```
This is a bit misleading, we are not actually restarting the container.
What do you think of "Could not resume tailing container %v" ?

mnshdw commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -159,13 +173,68 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
+		oldTailer, exists := l.tailers[containerID]
+		if exists {
+			source = oldTailer.source
+			oldTailer.Stop()
```
I'm not sure I understand why we stop the old one and create a new one.
Can you explain why we can't re-use it ?

achntrl replied:
There was an issue with a double close of a channel. I think @ajacquemot already had this issue in the udp tailer. We figured it was easier to create a new tailer and let the older get GCed 

```
2018-09-13 12:57:04 UTC | INFO | (tailer.go:93 in Start) | Start tailing container: da1e7badfbe8
2018-09-13 12:57:05 UTC | WARN | (launcher.go:219 in restartTailer) | Could not start tailer for container da1e7badfbe8: error during connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.25/containers/da1e7badfbe81904fbf40de8aa2629aaf648183e054e6db0c1d1e75e11c3e123/logs?follow=1&since=1536843424.788822639&stderr=1&stdout=1&tail=&timestamps=1: read unix @->/var/run/docker.sock: read: connection reset by peer
2018-09-13 12:57:06 UTC | INFO | (tailer.go:93 in Start) | Start tailing container: da1e7badfbe8
panic: close of closed channel

goroutine 228 [running]:
github.com/DataDog/datadog-agent/pkg/logs/decoder.(*Decoder).Stop(0xc4209eeed0)
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/decoder/decoder.go:96 +0x2e
github.com/DataDog/datadog-agent/pkg/logs/input/docker.(*Tailer).readForever(0xc420172b80)
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/input/docker/tailer.go:141 +0x4fb
created by github.com/DataDog/datadog-agent/pkg/logs/input/docker.(*Tailer).tail
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/input/docker/tailer.go:127 +0x13f
```

mnshdw commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -159,13 +173,68 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
+		// compute the offset to prevent from missing or duplicating logs
+		since, err := Since(l.registry, tailer.Identifier(), service.Before)
+		if err != nil {
+			log.Warnf("Could not recover tailing from last committed offset, start tailing from now for container %v: %v", ShortContainerID(containerID), err)
```
This could probably be simplified a bit too avoid being too verbose, maybe something along the lines of "Could not recover last committed offset for container %v" ?

mnshdw commented on pkg/logs/input/docker/tailer.go:
```diff
@@ -134,13 +137,21 @@ func (t *Tailer) readForever() {
+				default:
 					t.source.Status.Error(err)
-					log.Error("Err: ", err)
+					log.Errorf("Error reading logs for container %v: %v", ShortContainerID(t.ContainerID), err)
```
This is very nitpicky again, but I feel like our log messages are not very consistent both in vocabulary and grammar. It's probably not that important but I think having the same message pattern across a package helps with debugging and clarity. Here I've seen a bunch of "tailing" vs "reading logs", could we choose one and stick with it ? Same with "error" vs "could not".

mnshdw commented on pkg/logs/input/docker/tailer.go:
```diff
@@ -206,3 +217,10 @@ func (t *Tailer) checkForNewDockerTags() {
+// for more details, see: https://golang.org/src/internal/poll/fd.go#L18.
+func isClosedConnError(err error) bool {
+	return strings.Contains(err.Error(), "use of closed network connection")
+
```
nit: blank line

mnshdw commented on pkg/logs/input/docker/tailer.go:
```diff
@@ -31,8 +32,8 @@ const defaultSleepDuration = 1 * time.Second
-// With docker api, there is no way to know if a log comes from strout or stderr
-// so if we want to capture the severity, we need to tail both in two goroutines
+// Logs from stdout and stderr are multiplexed into a single channel and needs to be demultiplexed later one.
+// To multiplex logs, docker adds a header to all logs with format '[SEV][TS] [MSG]'.
```
👍

mnshdw commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -129,6 +141,8 @@ func (l *Launcher) run() {
 			l.stopTailer(containerID)
 			delete(l.pendingContainers, containerID)
+		case containerId := <-l.erroredContainerID:
+			go l.restartTailer(containerId)
```
What happens when we stop the agent and at least one restart routine is running ?

achntrl replied:
It stops: the launcher will receive a stop signal and transmit it to all the started tailers
The new tailer to start won't be able to `setupReader` and dies quickly. Even if it could it would receive the `isClosedConnError` and would exit

mnshdw reviewed (commented):
I left a lot of nitpicky comments about logging, but I have mostly two concerns: 
- Why don't we re-use old tailers ?
- Is having restart go-routines running going to create problems when stopping / restarting the agent ?

achntrl committed:
Wording changes

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -19,29 +22,38 @@ import (
+	cli                *client.Client
+	registry           auditor.Registry
+	stop               chan struct{}
+	erroredContainerID chan string
```
This name looks odd to me because the channel transport an error, would `containerToRestart` make more sense ?

achntrl replied:
No the channel transport the `ContainerID` to restart

ajacquemot replied:
Agreed, so why do we name it like so ? Could we find a better name please ?

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -159,13 +173,68 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
+func (l *Launcher) restartTailer(containerID string) {
+	log.Warnf("Tailing will restart for %v", ShortContainerID(containerID))
+	backoffDuration := backoffInitialDuration
+	cumulatedBackoffDuration := 0 * time.Second
```
nits, too verbose

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -159,13 +173,68 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
+		}
+
+		oldTailer, exists := l.tailers[containerID]
+		if exists {
```
I think this block could be moved outside of the for-loop

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -159,13 +173,68 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
+		tailer = NewTailer(l.cli, containerID, source, l.pipelineProvider.NextPipelineChan(), l.erroredContainerID)
+
+		// compute the offset to prevent from missing or duplicating logs
+		since, err := Since(l.registry, tailer.Identifier(), service.Before)
```
Same here

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -159,13 +173,68 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
+	var tailer *Tailer
+	var source *config.LogSource
+
+	for {
```
The for loop should only contain `err = tailer.Start(since) ...`

ajacquemot reviewed (commented):
This looks really good to me, left few comments.

achntrl committed:
Move creation of the new tailer outside of the loop

achntrl commented:
@mnshdw Regarding why we create a new tailer: there was an issue with a double close of a channel. I think @ajacquemot already had this issue in the udp tailer. We figured it was easier to create a new tailer and let the older get GCed 

```
2018-09-13 12:57:04 UTC | INFO | (tailer.go:93 in Start) | Start tailing container: da1e7badfbe8
2018-09-13 12:57:05 UTC | WARN | (launcher.go:219 in restartTailer) | Could not start tailer for container da1e7badfbe8: error during connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.25/containers/da1e7badfbe81904fbf40de8aa2629aaf648183e054e6db0c1d1e75e11c3e123/logs?follow=1&since=1536843424.788822639&stderr=1&stdout=1&tail=&timestamps=1: read unix @->/var/run/docker.sock: read: connection reset by peer
2018-09-13 12:57:06 UTC | INFO | (tailer.go:93 in Start) | Start tailing container: da1e7badfbe8
panic: close of closed channel

goroutine 228 [running]:
github.com/DataDog/datadog-agent/pkg/logs/decoder.(*Decoder).Stop(0xc4209eeed0)
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/decoder/decoder.go:96 +0x2e
github.com/DataDog/datadog-agent/pkg/logs/input/docker.(*Tailer).readForever(0xc420172b80)
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/input/docker/tailer.go:141 +0x4fb
created by github.com/DataDog/datadog-agent/pkg/logs/input/docker.(*Tailer).tail
	/home/vagrant/go/src/github.com/DataDog/datadog-agent/pkg/logs/input/docker/tailer.go:127 +0x13f
```

ajacquemot commented on pkg/logs/input/docker/launcher.go:
```diff
@@ -159,13 +173,67 @@ func (l *Launcher) startTailer(container *Container, source *config.LogSource) {
+		l.removeTailer(containerID)
+	}
+
+	tailer = NewTailer(l.cli, containerID, source, l.pipelineProvider.NextPipelineChan(), l.erroredContainerID)
```
nits:
```
tailer :=
```
and remove 
```
var tailer *Tailer
```

achntrl committed:
nits: change varaible declaration
